### PR TITLE
Add ARM32 SIMD bit, shift, boolean operations

### PR DESCRIPTION
**Pull request by gergocs**
Signed-off-by: Gergo Csizi gergocs@inf.u-szeged.hu

### CODE REVIEW

File: src/jit/SimdArm32Inl.h [hunk 1]
```diff
@@ -143,30 +160,108 @@ static void simdEmitI64x2GeS(sljit_compiler* compiler, sljit_s32 vd, sljit_s32 v
     simdEmitOp(compiler, SimdOp::vmvn | SimdOp::I32, vd, 0, vd);
 }
 
+static void simdEmitV128AnyTrue(sljit_compiler* compiler, sljit_s32 rd, sljit_s32 rm)
+{
+    auto tmpReg = SLJIT_FR2;
+
+    simdEmitOp(compiler, SimdOp::vpmax | SimdOp::I32 | SimdOp::unsignedBit, tmpReg, rm, getHighRegister(rm));
+    simdEmitOp(compiler, SimdOp::vpmax | SimdOp::I32 | SimdOp::unsignedBit, tmpReg, tmpReg, tmpReg);
```
Comment (zherczeg) [inline src/jit/SimdArm32Inl.h:168]: Does this needed?

File: src/jit/SimdArm32Inl.h [hunk 2]
```diff
@@ -143,30 +160,108 @@ static void simdEmitI64x2GeS(sljit_compiler* compiler, sljit_s32 vd, sljit_s32 v
...
+}
+
+static void simdEmitAllTrueEnd(sljit_compiler* compiler, sljit_s32 rd, sljit_s32 tmpReg, sljit_s32 type)
+{
+    sljit_emit_simd_lane_mov(compiler, SLJIT_SIMD_STORE | SLJIT_SIMD_REG_128 | SLJIT_SIMD_LANE_SIGNED | SLJIT_32 | type, tmpReg, 0, rd, 0);
+    sljit_emit_op2u(compiler, SLJIT_SUB | SLJIT_SET_Z, rd, 0, SLJIT_IMM, 0);
+    sljit_emit_op_flags(compiler, SLJIT_MOV, rd, 0, SLJIT_NOT_EQUAL);
+}
+
+static void simdEmitI8x16AllTrue(sljit_compiler* compiler, sljit_s32 rd, sljit_s32 rm)
```
Comment (zherczeg) [inline src/jit/SimdArm32Inl.h:181]: Could you generate these in some way with a loop?

File: src/jit/SimdArm32Inl.h [hunk 3]
```diff
@@ -199,16 +303,27 @@ static void emitUnarySIMD(sljit_compiler* compiler, Instruction* instr)
...
+    case ByteCode::V128NotOpcode:
+        simdEmitOp(compiler, SimdOp::Type::vmvn, dst, 0, args[0].arg);
+        break;
     default:
         ASSERT_NOT_REACHED();
         break;
     }
 
-    if (SLJIT_IS_MEM(args[1].arg)) {
+    if (!isDSTNormalRegister) {
```
Comment (zherczeg) [inline src/jit/SimdArm32Inl.h:323]: Something is wrong here. This is unnecessary if args[1] is a register

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/SimdArm32Inl.h (hunk 1 above)
Comment (gergocs) [reply src/jit/SimdArm32Inl.h:168]: yes it does.

File: src/jit/SimdArm32Inl.h (hunk 2 above)
Comment (gergocs) [reply src/jit/SimdArm32Inl.h:181]: yes

File: src/jit/SimdArm32Inl.h [hunk 4]
```diff
@@ -523,14 +663,196 @@ static void emitBinarySIMD(sljit_compiler* compiler, Instruction* instr)
...
+        } else {
+            simdEmitOp(compiler, SimdOp::Type::vneg | SimdOp::S8, dst, 0, dst);
+            simdEmitOp(compiler, SimdOp::Type::vshl | SimdOp::I8, dst, dst, args[0].arg);
+        }
+        break;
+    case ByteCode::I8X16ShrUOpcode:
+        if (SLJIT_IS_IMM(args[1].arg)) {
+            simdEmitOp(compiler, SimdOp::Type::vshrImm | SimdOp::SHL8 | (args[1].argw << 16) | (0b1 << 24), dst, 0, args[0].arg);
+        } else {
+            simdEmitOp(compiler, SimdOp::Type::vneg | SimdOp::S8, dst, 0, dst);
```
Comment (zherczeg) [inline src/jit/SimdArm32Inl.h:776]: Would it be a good idea to do the neg before `sljit_emit_op2(compiler, SLJIT_AND32, srcReg, 0, srcReg, 0, SLJIT_IMM, mask);`

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**